src/Interpreters/Cache/FileCacheFactory.cpp

```diff
@@ -48,4 +50,27 @@ FileCacheFactory::FileCacheData FileCacheFactory::getByName(const std::string &
     return *it->second;
 }
 
+void FileCacheFactory::updateSettingsFromConfig(const Poco::Util::AbstractConfiguration & config)
+{
+    CacheByName caches_by_name_copy;
```

Comment:
As a suggestion:
It seems that the update operation can be simplified with the help of the [compare-and-swap](https://en.wikipedia.org/wiki/Compare-and-swap) technique.

`applySettingsIfPossible` can be split into 'validate' and 'apply' parts, allowing `validate` to be performed without holding the cache mutex.
Please note that the new config should be treated as a whole, i.e. if one of its settings is invalid, ignore other settings.

Instead of adding `settings_mutex`, `FileCacheSettings` is defined as atomic: ```std::atomic<FileCacheSettings>```

```updateSettingsFromConfig```:

```cpp
FileCacheSettings new_settings;
new_settings.loadFromConfig(config, cache_info->config_path);

// Load the old settings into a local variable.
auto old_settings = settings.load();

// Validate the newly loaded settings, e.g., check the `background_download_queue_size_limit`.
if (!validateSettings(new_settings, old_settings))
  return;

// Check that the value stored in std::atomic is the same as the old value using https://en.cppreference.com/w/cpp/atomic/atomic/compare_exchange.
// If it has changed, return immediately since another thread has modified it.
if (!settings.compare_exchange_weak(old_settings, new_settings))
   return;

// `settings` store the `new_settings`.
// Apply the new_settings; applySettings acquires the cache mutex like it's presently done in `applySettingsIfPossible`.
cache_info->cache->applySettings(settings.load());
```

Wdyt?

Reply:
`applySettings` in the code above must be idempotent.

Reply:
> Instead of adding settings_mutex, FileCacheSettings is defined as atomic: std::atomic<FileCacheSettings>

Seems it is not possible. I am getting a compilation error: "error: _Atomic cannot be applied to type 'DB::FileCacheSettings' which is not trivially copyable".

Reply:
Okay, 'DB::FileCacheSettings contains a `std::string` which is not a POD-type.

Reply:
Okay, 'DB::FileCacheSettings contains a std::string which is not a POD-type.